base/bucket_gocb.go

```diff
@@ -1053,29 +1092,96 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
 
+	return bucket.deleteWithXattrInternal(k, xattrKey, nil)
+
+}
+
+// A function that will be called back after the first delete attempt but before second delete attempt
+type deleteWithXattrRaceInjection func(bucket CouchbaseBucketGoCB, k string, xattrKey string)
+
+// Delete doc + xattrs (purge).  Algorithm:
+// Try to delete body + xattr in single op
+// If fails
+// 		- Do get w/ xattr (get cas)
+// 		- Cas-safe delete just the xattr
+// If that fails with a cas error, return error (maybe race where someone resurrected doc)
+func (bucket CouchbaseBucketGoCB) deleteWithXattrInternal(k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
 	bucket.singleOps <- struct{}{}
 	defer func() {
 		<-bucket.singleOps
 	}()
 	gocbExpvars.Add("Delete", 1)
-	removeCas, err := bucket.Bucket.Remove(k, 0)
-	if err != nil && err != gocb.ErrKeyNotFound {
+
+	LogKeys["CRUD+"] = true
+
+	LogTo("CRUD+", "DeleteWithXattr called with key: %v xattrKey: %v", k, xattrKey)
+
+	// This flag seems to work well no matter what the current document state is in
+	deleteFlags := gocb.SubdocDocFlagAccessDeleted
+
+	// Try to delete body and xattrs in single op
+
+	// If the doc exists, delete both the doc body and the xattrs in one single op
+	_, mutateErr := bucket.Bucket.MutateInEx(k, deleteFlags, gocb.Cas(0), uint32(0)).
+		RemoveEx(xattrKey, gocb.SubdocFlagXattr). // Remove the xattr
+		RemoveEx("", gocb.SubdocFlagNone).        // Delete the document body
+		Execute()
+
+	// If no error, or it was just a ErrSubDocSuccessDeleted error, we're done.
+	// ErrSubDocSuccessDeleted is a "success error" that means "operation was on a tombstoned document"
+	if mutateErr == nil || mutateErr == gocbcore.ErrSubDocSuccessDeleted {
+		LogTo("CRUD+", "No error or ErrSubDocSuccessDeleted.  We're done.")
+		return nil
+	}
+
+	LogTo("CRUD+", "Initial attempt to delete doc failed with error: %v.  Trying again, this time only deleting doc body.", mutateErr)
+
+	// Invoke the callback which has the ability to change the document state
+	if callback != nil {
+		callback(
+			bucket,
+			k,
+			xattrKey,
+		)
+	}
+
+	//  Do get w/ xattr in order to get cas
+	var retrievedVal map[string]interface{}
+	var retrievedXattr map[string]interface{}
+	getCas, err := bucket.GetWithXattr(k, xattrKey, &retrievedVal, &retrievedXattr)
+	if err != nil {
 		return err
 	}
 
-	_, deleteXattrErr := bucket.Bucket.MutateInEx(k, gocb.SubdocDocFlagAccessDeleted, removeCas, 0).
-		RemoveEx(xattrKey, gocb.SubdocFlagXattr).
+	// If the doc body is non-empty at this point, then give up because it seems that a doc update has been
+	// interleaved with the purge.  Return error to the caller and cancel the purge.
+	if len(retrievedVal) != 0 {
+		return fmt.Errorf("DeleteWithXattr was unable to delete the doc.  It might be the case that another update " +
+			"was received which resurrected the doc by adding a new revision, in which case this delete operation is " +
+			"considred as cancelled.")
```

Comment:
I think this error message could be more assertive - I think we know what's going on here...the initial attempt to delete the document indicated that the document body didn't exist, but now we find it existing.  It's been updated while we attempted to purge, so we're cancelling the purge.  I think we can remove the 'might be the case' phrasing.